File: src/trx/game/ui/dialogs/photo_mode.c
#include <trx/game/ui/dialogs/photo_mode.h>

#include <trx/config.h>
#include <trx/game/game_strings/entries.h>
#include <trx/game/input.h>
#include <trx/game/lara/pose.h>
#include <trx/game/output/draw.h>
#include <trx/game/ui/elements/frame.h>
#include <trx/game/ui/elements/label.h>
#include <trx/game/ui/elements/modal.h>
#include <trx/game/ui/elements/pad.h>
#include <trx/game/ui/elements/spacer.h>
#include <trx/game/ui/elements/stack.h>
#include <trx/game/ui/scaler.h>

#include <stdio.h>

static bool M_HasIcon(const INPUT_ROLE role)
{
    const INPUT_BACKEND backend = g_Config.input.backend;
    const INPUT_LAYOUT layout = g_Config.input.layout[backend];
    return Input_GetKeyName(backend, layout, role, 0) != nullptr;
}

static void M_Title(const PHOTO_MODE current_mode)
{
    UI_BeginStackEx((UI_STACK_SETTINGS) {
        .orientation = UI_STACK_HORIZONTAL,
        .align = { .h = UI_STACK_H_ALIGN_DISTRIBUTE },
        .spacing = { .v = 8.0f },
    });
    switch (current_mode) {
    case PHOTO_MODE_CAMERA:
        UI_Label(GS("general/photo_mode/title_camera_pos"));
        break;
    case PHOTO_MODE_LARA_POS:
        UI_Label(GS("general/photo_mode/title_lara_pos"));
        break;
    }
    UI_Label("\\{input step_left}\\{input step_right}");
    UI_EndStack();
}

static void M_Inputs(const PHOTO_MODE current_mode)
{
    UI_Label(
        "\\{input camera_up}\\{input camera_down}"
        "\\{input camera_forward}\\{input camera_back}"
        "\\{input camera_left}\\{input camera_right}");
    UI_Label(
        "\\{input left}\\{input forward}"
        "\\{input back}\\{input right}");
    UI_Label("\\{input slow}+\\{input camera_up}/\\{input camera_down}");
    UI_Label("\\{input roll}");
    UI_Label("\\{input look}");

    UI_Label("[\\{input slow}+]\\{input draw}");
    if (Lara_Pose_IsAvailable()) {
        UI_Label("[\\{input slow}+]\\{input fly_cheat}");
    }
    UI_Label("[\\{input slow}+]\\{input pause}");
    UI_Label("\\{input toggle_ui}");
    UI_Label("\\{input action}");

    if (M_HasIcon(INPUT_ROLE_TOGGLE_PHOTO_MODE)
        && M_HasIcon(INPUT_ROLE_INVENTORY)) {
        UI_Label("\\{input toggle_photo_mode}/\\{input option}");
    } else if (M_HasIcon(INPUT_ROLE_TOGGLE_PHOTO_MODE)) {
        UI_Label("\\{input toggle_photo_mode}");
    } else if (M_HasIcon(INPUT_ROLE_INVENTORY)) {
        UI_Label("\\{input option}");
    }
}

static void M_Actions(const PHOTO_MODE current_mode)
{
    switch (current_mode) {
    case PHOTO_MODE_CAMERA:
        UI_Label(GS("general/photo_mode/camera_move_prompt"));
        UI_Label(GS("general/photo_mode/camera_rotate_prompt"));
        UI_Label(GS("general/photo_mode/camera_roll_prompt"));
        UI_Label(GS("general/photo_mode/camera_rotate_90_prompt"));
        UI_Label(GS("general/photo_mode/camera_reset_prompt"));
        break;
    case PHOTO_MODE_LARA_POS:
        UI_Label(GS("general/photo_mode/lara_move_prompt"));
        UI_Label(GS("general/photo_mode/lara_rotate_prompt"));
        UI_Label(GS("general/photo_mode/lara_roll_prompt"));
        UI_Label(GS("general/photo_mode/lara_rotate_90_prompt"));
        UI_Label(GS("general/photo_mode/lara_reset_prompt"));
        break;
    }
    UI_Label(GS("general/photo_mode/fov_prompt"));
    if (Lara_Pose_IsAvailable()) {
        UI_Label(GS("general/photo_mode/change_lara_pose"));
    }
    UI_Label(GS("general/photo_mode/advance_frame"));
    UI_Label(GS("general/photo_mode/toggle_help"));
    UI_Label(GS("general/photo_mode/snap_prompt"));
    UI_Label(GS("general/misc/exit"));
}

static void M_Body(const PHOTO_MODE current_mode)
{
    UI_BeginPad(8.0f, 8.0f);
    UI_BeginFrame(UI_FRAME_DIALOG_BACKGROUND);
    UI_BeginPad(8.0, 6.0);

    UI_BeginStackEx((UI_STACK_SETTINGS) {
        .orientation = UI_STACK_VERTICAL,
        .align = { .h = UI_STACK_H_ALIGN_SPAN },
        .spacing = { .v = 8.0f },
    });

    M_Title(current_mode);

    UI_BeginStackEx((UI_STACK_SETTINGS) {
        .orientation = UI_STACK_HORIZONTAL,
        .spacing = { .h = 8.0f },
    });

    // Inputs column
    UI_BeginStack(UI_STACK_VERTICAL);
    M_Inputs(current_mode);
    UI_EndStack();
    UI_BeginStack(UI_STACK_VERTICAL);
    M_Actions(current_mode);
    UI_EndStack();

    UI_EndStack();

    UI_EndStack();
    UI_EndPad();
    UI_EndFrame();
    UI_EndPad();
}

void UI_PhotoMode(const PHOTO_MODE current_mode)
{
    const int32_t frame_thickness =
        (int32_t)(UI_Scaler_Calc(4.0f, UI_SCALER_TARGET_TEXT) + 0.5f);
    Output_DrawPhotoModeFrame(frame_thickness);

    if (!g_Config.ui.enable_photo_mode_ui) {
        return;
    }

    float width = 0.0f;
    float height = 0.0f;
    UI_BeginMeasure();
    M_Body(current_mode);
    UI_EndMeasure(&width, &height);

    const float text_scale = UI_Scaler_GetTextScale();
    const float fit_scale =
        UI_GetFitScale(width / text_scale, height / text_scale);
    const float anchor = fit_scale < 1.0f ? 0.5f : 0.0f;

    UI_Scaler_PushTextScale(fit_scale);
    UI_BeginScreenModal(anchor, anchor);
    M_Body(current_mode);
    UI_EndModal();
    UI_Scaler_PopTextScale();
}

File: data/trx/ship/modules/overlay/photo_mode.lua
-- The photo mode help panel: what each key does while photo mode is open.
--
-- The panel is built once and updated through signals, like the rest of the
-- overlay. The engine draws the red border around the picture in the game view.

local input = trx.input
local ui = trx.ui
local signal = trx.signal

trx.locale.declare({
  ["general/photo_mode/title_camera_pos"] = "Photo Mode",
  ["general/photo_mode/title_lara_pos"] = "Moving Lara",
  ["general/photo_mode/camera_move_prompt"] = "Move camera",
  ["general/photo_mode/camera_rotate_prompt"] = "Rotate camera",
  ["general/photo_mode/camera_roll_prompt"] = "Roll camera",
  ["general/photo_mode/camera_rotate_90_prompt"] = "Rotate camera 90°",
  ["general/photo_mode/camera_reset_prompt"] = "Reset camera",
  ["general/photo_mode/lara_move_prompt"] = "Move Lara",
  ["general/photo_mode/lara_rotate_prompt"] = "Rotate Lara",
  ["general/photo_mode/lara_roll_prompt"] = "Roll Lara",
  ["general/photo_mode/lara_rotate_90_prompt"] = "Rotate Lara 90°",
  ["general/photo_mode/lara_reset_prompt"] = "Reset Lara",
  ["general/photo_mode/fov_prompt"] = "Adjust FOV",
  ["general/photo_mode/change_lara_pose"] = "Change Lara's pose",
  ["general/photo_mode/advance_frame"] = "Advance frame",
  ["general/photo_mode/toggle_help"] = "Toggle help",
  ["general/photo_mode/snap_prompt"] = "Take picture",
})

local state = {
  language = signal.config("language"),
  open = trx.game.signals.is_photo_mode,
}

state.shown = state.open & signal.config("ui.enable_photo_mode_ui")

state.target = signal.polled(function()
  return trx.game.photo_mode_target
end)

state.can_pose = signal.polled(function()
  return trx.lara.can_pose
end)

-- Show a key glyph when the player's device has one, and hide the prompt when
-- there is no binding to draw.
local function has_glyph(role)
  return state.language:map(function()
    return input.has_glyph(role)
  end)
end

local function text(shown, read)
  return ui.widgets.Label({ text = state.language:map(read), shown = shown })
end

-- A prompt follows the current target, so the same row can say "move the
-- camera" or "move Lara". Each line supplies its own translated text.
local function per_target(camera, lara)
  return signal.combine(state.target, state.language, function(target)
    if target == trx.game.PhotoModeTarget.CAMERA then
      return camera()
    end
    return lara()
  end)
end

-------------------------------------------------------------------------------
-- the title, and the keys that switch between camera and Lara
-------------------------------------------------------------------------------

local title = ui.widgets.Stack({
  orientation = ui.Orientation.HORIZONTAL,
  align = ui.HAlign.DISTRIBUTE,
  spacing = 8,
  children = {
    ui.widgets.Label({
      text = per_target(function()
        return trx.locale.get("general/photo_mode/title_camera_pos")
      end, function()
        return trx.locale.get("general/photo_mode/title_lara_pos")
      end),
    }),
    ui.widgets.Label({ text = "\\{input step_left}\\{input step_right}" }),
  },
})

-------------------------------------------------------------------------------
-- the keys, and what each of them does
-------------------------------------------------------------------------------

-- Name both exit keys when both are bound, or just the one that is available.
local exit_key = signal.combine(state.language, function()
  local photo = input.has_glyph(input.Role.TOGGLE_PHOTO_MODE)
  local option = input.has_glyph(input.Role.OPTION)
  if photo and option then
    return "\\{input toggle_photo_mode}/\\{input option}"
  elseif photo then
    return "\\{input toggle_photo_mode}"
  end
  return "\\{input option}"
end)

local exit_shown = signal.combine(
  has_glyph(input.Role.TOGGLE_PHOTO_MODE),
  has_glyph(input.Role.OPTION),
  function(photo, option)
    return photo or option
  end
)

local keys = ui.widgets.Stack({
  children = {
    ui.widgets.Label({
      text = "\\{input camera_up}\\{input camera_down}"
        .. "\\{input camera_forward}\\{input camera_back}"
        .. "\\{input camera_left}\\{input camera_right}",
    }),
    ui.widgets.Label({
      text = "\\{input left}\\{input forward}\\{input back}\\{input right}",
    }),
    ui.widgets.Label({
      text = "\\{input slow}+\\{input camera_up}/\\{input camera_down}",
    }),
    ui.widgets.Label({ text = "\\{input roll}" }),
    ui.widgets.Label({ text = "\\{input look}" }),
    ui.widgets.Label({ text = "[\\{input slow}+]\\{input draw}" }),
    ui.widgets.Label({
      text = "[\\{input slow}+]\\{input fly_cheat}",
      shown = state.can_pose,
    }),
    ui.widgets.Label({ text = "[\\{input slow}+]\\{input pause}" }),
    ui.widgets.Label({ text = "\\{input toggle_ui}" }),
    ui.widgets.Label({ text = "\\{input action}" }),
    ui.widgets.Label({ text = exit_key, shown = exit_shown }),
  },
})

local actions = ui.widgets.Stack({
  children = {
    ui.widgets.Label({
      text = per_target(function()
        return trx.locale.get("general/photo_mode/camera_move_prompt")
      end, function()
        return trx.locale.get("general/photo_mode/lara_move_prompt")
      end),
    }),
    ui.widgets.Label({
      text = per_target(function()
        return trx.locale.get("general/photo_mode/camera_rotate_prompt")
      end, function()
        return trx.locale.get("general/photo_mode/lara_rotate_prompt")
      end),
    }),
    ui.widgets.Label({
      text = per_target(function()
        return trx.locale.get("general/photo_mode/camera_roll_prompt")
      end, function()
        return trx.locale.get("general/photo_mode/lara_roll_prompt")
      end),
    }),
    ui.widgets.Label({
      text = per_target(function()
        return trx.locale.get("general/photo_mode/camera_rotate_90_prompt")
      end, function()
        return trx.locale.get("general/photo_mode/lara_rotate_90_prompt")
      end),
    }),
    ui.widgets.Label({
      text = per_target(function()
        return trx.locale.get("general/photo_mode/camera_reset_prompt")
      end, function()
        return trx.locale.get("general/photo_mode/lara_reset_prompt")
      end),
    }),
    text(nil, function()
      return trx.locale.get("general/photo_mode/fov_prompt")
    end),
    text(state.can_pose, function()
      return trx.locale.get("general/photo_mode/change_lara_pose")
    end),
    text(nil, function()
      return trx.locale.get("general/photo_mode/advance_frame")
    end),
    text(nil, function()
      return trx.locale.get("general/photo_mode/toggle_help")
    end),
    text(nil, function()
      return trx.locale.get("general/photo_mode/snap_prompt")
    end),
    text(nil, function()
      return trx.locale.get("general/misc/exit")
    end),
  },
})

-------------------------------------------------------------------------------
-- the panel itself
-------------------------------------------------------------------------------

-- Keep the panel off the edge of the screen, so the frame only needs to carry
-- the margin between its edge and the text.
local panel = ui.widgets.Fit({
  shown = state.shown,
  child = ui.widgets.Frame({
    style = ui.FrameStyle.DIALOG,
    child = ui.widgets.Pad({
      x = 8,
      y = 6,
      child = ui.widgets.Stack({
        spacing = 8,
        align = ui.HAlign.SPAN,
        children = {
          title,
          ui.widgets.Stack({
            orientation = ui.Orientation.HORIZONTAL,
            spacing = 8,
            children = { keys, actions },
          }),
        },
      }),
    }),
  }),
})

ui.regions.place(ui.Region.TOP_LEFT, panel)
